### ISSUE TITLE
Get `411 Length Required` error when using JSON as request body

### ISSUE THREAD

**Issue by AaronJan** [guzzle/guzzle#1645]
Well, I ran into this problem today, finally I found out it's caused by Tengine (version `1.5.6`, based on Nginx `1.2.9`).

Here is my code:
```php
<?php

$config = ['base_uri' => 'http://my.project'];
$client = new \GuzzleHttp\Client($config);

$response = $client->request('GET', '/get-info', [
    'json' => ['key' => 'value'],
]);

// ...

```

The error response looks like this:

```html
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0/EN">
<html>
<head><title>411 Length Required</title></head>
<body> .......
```

I found 2 ways to bypass this error:

* Use `POST` instead of `GET`
* Add `Content-Length` header (for example: `Content-Length: 0`, the value doesn't affect anything)

There're some people talk about the second method, but since I updated my Tengine, the problem is gone.

My question is: should Guzzle set the `Content-Length` header by default when using JSON?

I haven't found any RFC about this, but in [here](http://www.json.org/JSONRequest.html), every example have a `Content-Length` header.

**Comment by AaronJan** [comment]
If so, I'd like to make the PR. 😄

**Comment by sagikazarmark** [comment]
Well, I think the proper solution would be to use a middleware, like here: https://github.com/php-http/client-common/blob/master/src/Plugin/ContentLengthPlugin.php

Although it requires the stream to be readable.

**Comment by sagikazarmark** [comment]
Oh, there is already a middleware for it: https://github.com/guzzle/guzzle/blob/master/src/PrepareBodyMiddleware.php

It's also added to the handler stack by default: https://github.com/guzzle/guzzle/blob/master/src/HandlerStack.php#L44

So I wonder what the problem is.

**Comment by AaronJan** [comment]
I'm so sorry for your time, after read your reply I did some tests, it did have the `Content-Length` header by default, by when I did this test on that server which has the `411 error`, the `Content-Length` header is missing. Until now I found out it's a Nginx configuration problem.

I was careless that I didn't do the test on other server environment. Thank you very much! 👍 

I really don't know much about `Handler & Middleware`, I'll catch up with that.

By the way, a little more trouble, is there any way to get the final URL and headers that I send?

```php
<?php

$config = ['base_uri' => 'http://my.project'];
$client = new \GuzzleHttp\Client($config);

$response = $client->request('GET', '/get-info', [
    'json' => ['key' => 'value'],
]);

// In here
```
